x/superfluid/keeper/slash.go

```diff
@@ -59,11 +63,72 @@ func (k Keeper) SlashLockupsForValidatorSlash(ctx sdk.Context, valAddr sdk.ValAd
 func (k Keeper) slashSynthLock(ctx sdk.Context, synthLock *lockuptypes.SyntheticLock, slashFactor sdk.Dec) {
 	// Only single token lock is allowed here
 	lock, _ := k.lk.GetLockByID(ctx, synthLock.UnderlyingLockId)
-	slashAmt := lock.Coins[0].Amount.ToDec().Mul(slashFactor).TruncateInt()
-	slashCoins := sdk.NewCoins(sdk.NewCoin(lock.Coins[0].Denom, slashAmt))
+	slashAmt := lock.Coins[0].Amount.ToDec().Mul(slashFactor)
+	lockSharesToSlash := sdk.NewCoins(sdk.NewCoin(lock.Coins[0].Denom, slashAmt.TruncateInt()))
+
+	// If the slashCoins contains a cl denom, we need to update the underlying cl position to reflect the slash.
 	_ = osmoutils.ApplyFuncIfNoError(ctx, func(cacheCtx sdk.Context) error {
-		// These tokens get moved to the community pool.
-		_, err := k.lk.SlashTokensFromLockByID(cacheCtx, lock.ID, slashCoins)
-		return err
+		if strings.HasPrefix(lock.Coins[0].Denom, cltypes.ClTokenPrefix) {
+			// Run prepare logic to get the underlying coins to slash.
+			// We get the pool address here since the underlying coins will be sent directly from the pool to the community pool instead of the lock module account.
+			// Additionally, we update the cl position's state entry to reflect the slash in the position's liquidity.
+			poolAddress, underlyingCoinsToSlash, err := k.prepareConcentratedLockForSlash(cacheCtx, lock, slashAmt)
+			if err != nil {
+				return err
+			}
+			// Run the normal slashing logic, but instead of sending gamm shares to the community pool, we send the underlying coins
+			// the cl shares represent to the community pool and burn the cl shares from the lockup module account as well as the lock itself
+			_, err = k.lk.SlashTokensFromLockByIDSendUnderlyingAndBurn(cacheCtx, lock.ID, lockSharesToSlash, underlyingCoinsToSlash, poolAddress)
+			return err
+		} else {
+			// These tokens get moved to the community pool.
+			_, err := k.lk.SlashTokensFromLockByID(cacheCtx, lock.ID, lockSharesToSlash)
+			return err
+		}
 	})
 }
+
+// prepareConcentratedLockForSlash is a helper function that runs pre-slash logic for concentrated lockups. This function:
+// 1. Figures out the underlying assets from the liquidity being slashed and creates a coin object this represents
+// 2. Sets the cl position's liquidity state entry to reflect the slash
+// 3. Returns the pool address that will sends the underlying coins as well as the underlying coins to slash
+func (k Keeper) prepareConcentratedLockForSlash(ctx sdk.Context, lock *lockuptypes.PeriodLock, slashAmt sdk.Dec) (sdk.AccAddress, sdk.Coins, error) {
+	// Get the position ID from the lock denom
+	positionID, err := cltypes.GetPositionIdFromShareDenom(lock.Coins[0].Denom)
```

Comment:
Let's do a length check just in case

Reply:
what are we length checking? I don't see anything here that is static in terms of its length.

Reply:
Added coins length check here https://github.com/osmosis-labs/osmosis/pull/4884/commits/8720afc77202343b8fdd921e1d34eb1351a32ef0